### ISSUE TITLE
v7.0 rewrite checklist

### ISSUE THREAD

**Issue by yamalight** [exoframejs/exoframe#311]
This is an issue to track overall progress on v7.0 rewrite.
Plan is to rewrite / update code to use latest libs (not just versions) and ES modules.
Additionally, I'm planning to implement `exoframe-client` library that provides simple API for interacting with server (would allow to create new client, e.g. dashboard as some people wanted).

Packages:

- [ ] exoframe-client
- [ ] exoframe-server
- [ ] exoframe-faas
- [ ] exoframe-cli
- [ ] exoframe-website (incl. docs)
- [ ] exoframe-recipe-ghost
- [ ] exoframe-template-java

Planned major changes:

- [ ] Drop docker swarm support. 
       Reasoning: support for it adds complexity, almost nobody uses swarm (k8s is a better solution for that problem), exoframe doesn't work well in those scenarios as a whole.
- [ ] Drop docker-compose support.
       Reasoning: compose-based deployments don't work well with exoframe (users has to manually add traefik labels, etc), which causes a lot of issues. there's also an issue with tearing down and rebuilding / restarting all services when you only want to update one. IMO the best way would be to deprecate it and describe how to deploy multi-service projects using plain exoframe configs.
- [ ] Drop plugins support
       Reasoning: swarm is deprecated. There are no other plugins and doesn't seem like they'd be useful 🤔
- [ ] Drop old / unused recipes and templates (include one of each as an example)
       Reasoning: it's hard to maintain / fix all of them with time (I kinda ended up abandoning most of them), so it's better to just have a few simple examples.

Planned issues:
- [ ] Use XDG_CONFIG_HOME for config location (#309)
- [ ] Traefik dashboard setting (#147)
- [ ] More private key types (#191)
- [ ] Removal of multiple deployments (#291)
- [ ] Docker images hosted on github (#302)
- [ ] New website (#259)
- [ ] Docs for new traefik config (#139, #303)
- [ ] Docs for typical deployment cases (#235)
- [ ] Docs for usage in CI / github action (#260)

"If there's time" issues:
- [ ] Analytics (#270)
- [ ] Deterministic deployment IDs (#275)
- [ ] Custom error pages (#305)

ToDos:
- [ ] Combine all repos to monorepo
- [ ] Archive all old repos

Any feedback welcome.

**Comment by yamalight** [comment]
Apparently, jest currently doesn't support ES module mocking (see https://github.com/facebook/jest/issues/10025), so server rewrite is blocked on that feature.

**Comment by yamalight** [comment]
I wanted to use Ink for new cli, but it seems like it also has issues with ES modules support (`import-jsx` package specifically). So CLI is also blocked 🙃

**Comment by yamalight** [comment]
Great collection of [CLI websites](https://news.ycombinator.com/item?id=26878936) on HN.

**Comment by yamalight** [comment]
Alright, few months later - doesn't seem like native ESM support in Node.js ecosystem is going to happen anytime soon. Will probably scrap the plan to use pure ESM and use babel (with e.g. rollup) to compile to cjs. 
Well, was worth a try I guess 🙃

**Comment by yamalight** [comment]
I'm starting to think it might be easier to rewrite whole thing to typescript at this stage. So many tools are still not quite there yet with ESM support. The pain 🙃

**Comment by FDiskas** [comment]
Yes please 😊

**Comment by yamalight** [comment]
I've tried doing that yesterday and as just a solution to ESM issues - it's no longer viable. There already are ESM-only modules that won't work in CJS envs, so you have to compile ts to ESM. Which just means you have ts on top of all existing issues 😅

**Comment by FDiskas** [comment]
I can try to make my hands dirty 😉

**Comment by yamalight** [comment]
@FDiskas I'd prefer to first finish ESM version and then do anything else. There's too many issues as it is to add yet another thing on top 😅

**Comment by niieani** [comment]
Hi @yamalight! I'm really sad to see "Drop docker-compose support." on the list. This is basically how 90% of my exoframe deployments are done. 

I really love this feature and it would be really sad to see it go, as I'll have to either keep using the old version or create a fork. At this point I'd be willing to even contribute a small amount of cash to support for docker-compose, or help by contributing to maintaining that feature myself.

I've [previously contributed a PR](https://github.com/exoframejs/exoframe-server/pull/16) related to this feature, and as you can see I've been using exoframe for over 5 years now. 

I really don't mind the fact that I need to specify labels manually, in fact, I prefer it, because it gives me more flexibility as to how the deployments should be routed via traefik.

I really hope you reconsider. Let me know if you'd like to know more about how my use cases. Using exoframe over the years has been fantastic!

**Comment by yamalight** [comment]
@niieani I still think it's one of the worst ways to deploy things via exoframe - you make your server tear-down, re-build and re-start _all containers_ defined in the compose file on __every deployment__. it's incredibly wasteful in all sorts of ways. vast majority of the time you want to update just one (app) container, and keep all the others unchanged (database, message queue, etc).

this is the main reason I think "exoframe way" (tm) is just a better approach.
and it's not like doing it "exoframe way" is a lot of work - it takes a few minutes more than deploying compose, but only _once_.

here's an example. Imagine you have following compose file:
```yaml
version: '3.6'
services:
  postgres:
    image: postgres:15
    ports:
      - 5432:5432
    volumes:
      - db_data:/var/lib/postgresql/data
    environment:
      POSTGRES_PASSWORD: postgrespassword
  redis:
    image: redis:7
    ports:
      - '6379:6379'
  app:
    build: .
    ports:
      - 8080:8080
    depends_on:
      - postgres
      - redis
    environment:
      DATABASE_URL: postgres://postgres:postgrespassword@postgres:5432/postgres
      REDIS_HOST: redis

volumes:
  db_data:
```

This work perfectly fine for local testing / dev, but deploying it via exoframe 6 will tear-down and re-start postgres and redis _every_ time you deploy. Why waste that time?

Here's how you would deploy it using exoframe configs:
First, you define config for postgres:
```json
{
  "name": "my-postgres",
  "domain": false,
  "image": "postgres:15",
  "volumes": ["db_data:/var/lib/postgresql/data"],
  "hostname": "mypostgres",
  "env": {
    "POSTGRES_PASSWORD": "postgrespassword"
  }
}
```
Next, you define exoframe config for redis:
```json
{
  "name": "my-redis",
  "domain": false,
  "image": "redis:7",
  "hostname": "myredis"
}
```
And finally, you define config for your app:
```json
{
  "name": "my-app",
  "domain": "my-app.com",
  "env": {
    "DATABASE_URL": "postgres://postgres:postgrespassword@postgres:5432/postgres",
    "REDIS_HOST": "askaredis"
  }
}
```

That's it, that's all the work it takes and you do it _once_.
After doing that - you can individually (re-)deploy any of the containers without touching the other ones.

I think it's a fairly simple procedure that does make deployments _a lot_ faster / simpler.
It obviously has to be documented a lot better than it is right now - but that should come along with a new website.

I was also thinking about a possibility of writing a CLI script that would auto-convert compose files into sub-deployments like this.

Would love to know your thoughts on this - is that easy enough? Would you still want compose?

**Comment by FDiskas** [comment]
@yamalight hi, I noticed that repository with a website is archived. Do you need a help on building the new one? Thinking about astrojs

**Comment by yamalight** [comment]
@FDiskas website will also be part of current monorepo (old repo was folded into it as-is). 
appreciate the offer, but haven't gotten around to that part yet! 😅
